pkg/controller/atlasdeployment/advanced_deployment.go

```diff
@@ -1,333 +1,128 @@
-			}
+		result = EnsureCustomZoneMapping(
+			ctx,
+			deploymentInAKO.GetProjectID(),
+			deploymentInAKO.GetCustomResource().Spec.DeploymentSpec.CustomZoneMapping,
+			deploymentInAKO.GetName(),
+		)
+		if transition = r.transitionFromResult(ctx, deploymentInAKO.GetProjectID(), deploymentInAKO.GetCustomResource(), result); transition != nil {
+			return transition(workflow.Internal)
 		}
-	}
-
-	return atlas
-}
 
-// GetAllDeploymentNames returns all deployment names including regular and advanced deployment.
-func GetAllDeploymentNames(ctx context.Context, client *mongodbatlas.Client, projectID string) ([]string, error) {
-	var deploymentNames []string
+		result = EnsureManagedNamespaces(
+			ctx,
+			deploymentInAKO.GetProjectID(),
+			deploymentInAKO.ClusterType,
+			deploymentInAKO.GetCustomResource().Spec.DeploymentSpec.ManagedNamespaces,
+			deploymentInAKO.GetName(),
+		)
+		if transition = r.transitionFromResult(ctx, deploymentInAKO.GetProjectID(), deploymentInAKO.GetCustomResource(), result); transition != nil {
+			return transition(workflow.Internal)
+		}
 
-	advancedDeployments, _, err := client.AdvancedClusters.List(ctx, projectID, &mongodbatlas.ListOptions{})
-	if err != nil {
-		return nil, err
-	}
+		err = customresource.ApplyLastConfigApplied(ctx.Context, deploymentInAKO.GetCustomResource(), r.Client)
+		if err != nil {
+			return r.terminate(ctx, workflow.Internal, err)
+		}
 
-	for _, d := range advancedDeployments.Results {
-		deploymentNames = append(deploymentNames, d.Name)
+		return r.ready(ctx, deploymentInAKO.GetCustomResource(), deploymentInAtlas)
+	case status.StateCREATING:
+		return r.inProgress(ctx, deploymentInAKO.GetCustomResource(), deploymentInAtlas, workflow.DeploymentCreating, "deployment is provisioning")
+	case status.StateUPDATING, status.StateREPAIRING:
+		return r.inProgress(ctx, deploymentInAKO.GetCustomResource(), deploymentInAtlas, workflow.DeploymentUpdating, "deployment is updating")
+	case status.StateDELETING, status.StateDELETED:
+		return workflow.OK().ReconcileResult(), nil
+	default:
+		return r.terminate(ctx, workflow.Internal, fmt.Errorf("unknown deployment state: %s", deploymentInAtlas.GetState()))
 	}
-
-	return deploymentNames, nil
 }
 
-func (r *AtlasDeploymentReconciler) ensureConnectionSecrets(ctx *workflow.Context, project *akov2.AtlasProject, name string, connectionStrings *mongodbatlas.ConnectionStrings, deploymentResource *akov2.AtlasDeployment) workflow.Result {
+func (r *AtlasDeploymentReconciler) ensureConnectionSecrets(ctx *workflow.Context, deploymentInAKO deployment.Deployment, connection *status.ConnectionStrings) error {
 	databaseUsers := akov2.AtlasDatabaseUserList{}
 	err := r.Client.List(ctx.Context, &databaseUsers, &client.ListOptions{})
 	if err != nil {
-		return workflow.Terminate(workflow.Internal, err.Error())
+		return err
 	}
 
+	atlasDeployment := deploymentInAKO.GetCustomResource()
 	secrets := make([]string, 0)
 	for i := range databaseUsers.Items {
```

Comment:
This loop doesn't seem to have test coverage - is it worth adding a database user to a test so we can test this?

Reply:
good point! the reason(s) I didn't add a test to cover these changes were two: 1) connection secrets are extensively tested on integration tests, which is a good safenet. 2) There a proposal to extract connection secret handling into a separate component or even a controller.